### ISSUE TITLE
Register new user under different Tenant (Tenant impersonation)

### ISSUE THREAD

**Issue by coommark** [Finbuckle/Finbuckle.MultiTenant#347]
My application is built using .NET Core 3.1 with Finbuckle.MultiTenant.AspNetCore 6.0.0.

As the super administrator, I want to create an administrator for another new tenant. Note that

1. I am able to set tenant using TrySetTenantInfo()
2. However when I register a new user, the user is registered under my current logged in tenant. So in the AspNetUsers table, the Tenant Id of the new user is the Id of the tenant I am logged in with.

Please can someone point me in the right direction?

My code is:

```
           //Correctly gets the current tenant
            var currentTenant = HttpContext.GetMultiTenantContext<ApplicationTenantInfo>()?.TenantInfo;
            
            //Correctly retrieves the target tenant
            var targetTenant = _store.TryGetAsync(model.TenantId.ToString()).Result;

            //Correctly sets the new tenant to the target tenant
            bool isSet = HttpContext.TrySetTenantInfo(targetTenant, true);

            //Verify that target tenant is correctly set, works
            var newTenant = HttpContext.GetMultiTenantContext<ApplicationTenantInfo>()?.TenantInfo;       

            if (ModelState.IsValid)
            {
                var password = RandomGenerator.GenerateRandomString(8);
                var user = new ApplicationUser { UserName = model.Email, Email = model.Email, Name = model.Name, PhoneNumber = model.PhoneNumber };                
                
                //User is not created using the target tenant
                var result = await _userManager.CreateAsync(user, password);
                //bool reset = HttpContext.TrySetTenantInfo(currentTenant, false);
            }

            var tenants = _store.GetAllAsync().Result;
            ViewBag.TenantId = new SelectList(tenants, "Id", "Identifier");
            return View(model);
        }
```

I have also tried:

```
var accessor = _serviceProvider.GetRequiredService<IMultiTenantContextAccessor>();
            var multiTenantContext = new MultiTenantContext<ApplicationTenantInfo>();
            multiTenantContext.TenantInfo = newTenant;
            accessor.MultiTenantContext = multiTenantContext;
```

**Comment by AndrewTriesToCode** [comment]
hi @coommark 

You are getting into some more advanced stuff-- very cool. I think the issue is that the user manager might be holding onto the old db context with the original tenant internally. After you call `TrySetTenantInfo` with the `true` 2nd parameter, try re-resolving the user manager from DI (e.g. from `HttpRequest.RequestServices`.

Let me know if that still doesn't work and I'll dig deeper for you.

-Andrew

**Comment by coommark** [comment]
Hello @AndrewTriesToCode, thank you a MILLION for the response! I tried the following as per your suggestions:

**This did not work**
```
            bool isSet = HttpContext.TrySetTenantInfo(targetTenant, true);

            //Try re-resolving the user manager from DI as per @AndrewTriesToCode          
            var userManager = HttpContext.RequestServices.GetService<UserManager<ApplicationUser>>();
            
            if (ModelState.IsValid)
            {
                var password = RandomGenerator.GenerateRandomString(8);
                var user = new ApplicationUser { UserName = model.Email, Email = model.Email, Name = model.Name, PhoneNumber = model.PhoneNumber };                
                
                //User is not created using the target tenant
                var result = await _userManager.CreateAsync(user, password);
                if(result.Succeeded)
                {
                    //Send email with password, tenant name, login link                    
                }

                bool reset = HttpContext.TrySetTenantInfo(currentTenant, true);
            }
```

**This also did not work**
```
           bool isSet = HttpContext.TrySetTenantInfo(targetTenant, true);

            //Try re-resolving the user manager from DI as per @AndrewTriesToCode            
            var scope = _serviceProvider.CreateScope();            
            var newContext = HttpContext.RequestServices = scope.ServiceProvider;
            var userManager = newContext.GetService<UserManager<ApplicationUser>>();
            
            if (ModelState.IsValid)
            {
                var password = RandomGenerator.GenerateRandomString(8);
                var user = new ApplicationUser { UserName = model.Email, Email = model.Email, Name = model.Name, PhoneNumber = model.PhoneNumber };                
                
                //User is not created using the target tenant
                var result = await _userManager.CreateAsync(user, password);
                if(result.Succeeded)
                {
                    //Send email with password, tenant name, login link                    
                }

                bool reset = HttpContext.TrySetTenantInfo(currentTenant, true);
            }
```

Please any help?

**Comment by AndrewTriesToCode** [comment]
Hi ok I’ll start digging in. Do you have a simple project or repo that shows the problem you can share with me?

**Comment by coommark** [comment]
Hello @AndrewTriesToCode sorry it took a while, I had to create a smaller application because the other one is too large with lots of other stuff. Interestingly even the small app still doesn't work. Below is the link to the repo:

[https://github.com/coommark/finbuckle_test/tree/master](https://github.com/coommark/finbuckle_test/tree/master)

The instructions are there. When you login and go to `https://localhost:44351/pgl/Admin/Create` the tenants will be loaded on the view below and you can test.

Thank you plenty!
![Screenshot (52)](https://user-images.githubusercontent.com/5004140/98644707-2e2cf500-2331-11eb-84a8-9b08c361e24d.png)

**Comment by AndrewTriesToCode** [comment]
Thanks @coommark 

I cloned the repo. Interestingly GitHub seemed to have changed the default branch to `main` instead of `master` so I was confused for a moment... Also I had to change the database to SQLite since I use a Mac, but I got it running. I'm digging into it as we speak.

**Comment by AndrewTriesToCode** [comment]
@coommark I think I've got it and if I'm right you might want to kick yourself (I say that as someone who does this to my self frequently):

In your first attempt above:
```cs
            bool isSet = HttpContext.TrySetTenantInfo(targetTenant, true);

            //Try re-resolving the user manager from DI as per @AndrewTriesToCode          
            var userManager = HttpContext.RequestServices.GetService<UserManager<ApplicationUser>>();
            
            if (ModelState.IsValid)
            {
                var password = RandomGenerator.GenerateRandomString(8);
                var user = new ApplicationUser { UserName = model.Email, Email = model.Email, Name = model.Name, PhoneNumber = model.PhoneNumber };                
                
                
                if(result.Succeeded)
                {
                    //Send email with password, tenant name, login link                    
                }

                bool reset = HttpContext.TrySetTenantInfo(currentTenant, true);
            }
```

Note that this line is not using the newly resolved `userManager` but instead the original `_userManager`:
```cs
//User is not created using the target tenant
var result = await _userManager.CreateAsync(user, password);
```

Let me know how it goes!

**Comment by coommark** [comment]
@AndrewTriesToCode Aha! Just now I used more expletives in one sentence than a sailor would in a day - right before I kicked myself twice as you recommended. You got it, that's exactly what happened! Thank you so very much for pointing me in the direction of re-resolving the user manager as per your first comment. 

This was a good experience and forced me to read some of the very well-written code of this awesome multi-tenant framework.  I've used some of the most popular, even paid ones. This gives me just what I need - handling the multitenant part of the application without all the usual baggage associated with those other frameworks. Thank you for this great framework.
